data_structures/doubly_linked_list.js

```diff
@@ -48,37 +71,218 @@ class Node {
  */
 
 class DoublyLinkedList {
-    constructor(){}
+    constructor(){
+        this.head = null
+        this.tail = null
+    }
     
     //add a new node with the given value to the tail of the list
-    add_to_tail(value){}
+    add_to_tail(value){
+        let new_node = new Node(value)
+
+        if(!this.tail){
+            this.head = new_node
+            this.tail = new_node
+            
+            return this
+        }
+
+        let current_tail = this.tail
+        new_node.prev = current_tail
+        current_tail.next = new_node
+        this.tail = new_node
+
+        return this
+    }
     
     //returns the value of the removed node or null if list is empty
-    remove_tail(){}
+    remove_tail(){
+        if(!this.tail){
+            return null
+        }
+
+        let tail = this.tail
+
+        if(this.head === this.tail){
+            this.head = null
+            this.tail = null
+
+            return tail.value
+        }
+
+        this.tail = tail.prev
+        this.tail.next = null
+
+        return tail.value
+    }
     
     //add a new node with the given value to the head of the list
-    add_to_head(value){}
+    add_to_head(value){
+        let new_node = new Node(value)
+
+        if(!this.head){
+            this.head = new_node
+            this.tail = new_node
+
+            return this
+        }
+
+        let current_head = this.head
```

Comment:
A different way of doing this and not saving this.head to a variable would be to go in the order of 
1. set the new node next pointer to the current head
2. set the current head's prev pointer to the new node
3. set the list head's reference to the new node

So it would look like
```
    new_node.next = this.head;
      this.head.prev = new_node;
      this.head = new_node;```